### PR TITLE
Issue/4408 cache invalidation

### LINKED ISSUE #4408: Update cache invalidation mechanism for cfc files
The current caching mechanism for `.cf` files uses the modification date of the `.cf` file and the `.cfc` file to determine whether a cache file is still valid. Since #4407 stores these cache files in the root of the inmanta project, this check is not longer sufficient. The modification timestamp of the file being cached should be stored in `.cfc` file to determine whether the `.cfc` file is still valid.

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

More precise  invalidation for the compiler cache
closes #4408 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design

### CODE REVIEW

File: src/inmanta/parser/cache.py [hunk 1]
```diff
@@ -27,6 +27,12 @@
 LOGGER = logging.getLogger(__name__)
 
 
+class CacheEnveloppe:
```
Comment (sanderr) [inline src/inmanta/parser/cache.py:30]:
```suggestion
class CacheEnvelope:
```
Two p's is the Dutch word.

File: src/inmanta/parser/cache.py [hunk 2]
```diff
@@ -74,13 +80,22 @@ def un_cache(self, namespace: Namespace, filename: str) -> Optional[List[Stateme
             if not os.path.exists(cache_filename):
                 self.misses += 1
                 return None
+            mtime = os.path.getmtime(filename)
             if os.path.getmtime(filename) > os.path.getmtime(cache_filename):
                 self.misses += 1
                 return None
```
Comment (sanderr) [inline src/inmanta/parser/cache.py:86]: Is this still relevant? I think we can just consider the envelope timestamp the authority and ignore the cache mtime altogether.

Comment (sanderr) [reply src/inmanta/parser/cache.py:86]: Or is this meant as a performance boost in the most common scenario so we don't need to read the file?

File: tests/compiler/test_parser_cache.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
+"""
+    Copyright 2018 Inmanta
```
Comment (sanderr) [inline tests/compiler/test_parser_cache.py:2]:
```suggestion
    Copyright 2022 Inmanta
```

File: tests/compiler/test_parser_cache.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+def test_caching(snippetcompiler):
+    # reset counts
+    parser.cache_manager = CacheManager()
+    snippetcompiler.setup_for_snippet(
+        """
+a=1
+"""
+    )
+    # don't know hit count, may very on previous testcases
```
Comment (sanderr) [inline tests/compiler/test_parser_cache.py:34]:
```suggestion
    # don't know hit count, may vary on previous testcases
```
Might not be what you meant. If not, I don't understand.

File: src/inmanta/parser/cache.py (hunk 1 above)
Comment (wouterdb) [reply src/inmanta/parser/cache.py]: done

File: src/inmanta/parser/cache.py [hunk 5]
```diff
@@ -27,6 +27,14 @@
 LOGGER = logging.getLogger(__name__)
 
 
+class CacheEnvelope:
+    """Every cached file gets the exact modification to of the file it is caching, to have cheap, accurate invalidation"""
```
Comment (arnaudsjs) [inline src/inmanta/parser/cache.py:31]:
```suggestion
    """Every cached file gets the exact modification time of the file it is caching, to have cheap, accurate invalidation"""
```

File: src/inmanta/parser/cache.py [hunk 6]
```diff
@@ -27,6 +27,14 @@
...
+
+    def __init__(self, timestamp: float, statements: List[Statement]):
```
Comment (arnaudsjs) [inline src/inmanta/parser/cache.py:33]:
```suggestion
    def __init__(self, timestamp: float, statements: List[Statement]) -> None:
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

**Review by arnaudsjs — APPROVED** [review]

**Review by sanderr — APPROVED** [review]

**Pull request closed without merging**